File: 1438-longest-continuous-subarray-with-absolute-diff-less-than-or-equal-to-limit/1438-longest-continuous-subarray-with-absolute-diff-less-than-or-equal-to-limit.rb
# @param {Integer[]} nums
# @param {Integer} limit
# @return {Integer}
def longest_subarray(nums, limit)
  max_deque = []
  min_deque = []
  left = 0
  result = 0

  nums.each_with_index do |num, right|
    # Maintain the max_deque to have decreasing elements
    while !max_deque.empty? && max_deque[-1] < num
      max_deque.pop
    end
    max_deque.push(num)

    # Maintain the min_deque to have increasing elements
    while !min_deque.empty? && min_deque[-1] > num
      min_deque.pop
    end
    min_deque.push(num)

    # If the current window is invalid, move the left pointer to shrink the window
    while max_deque[0] - min_deque[0] > limit
      left += 1
      max_deque.shift if max_deque[0] == nums[left - 1]
      min_deque.shift if min_deque[0] == nums[left - 1]
    end

    # Update the result for the valid window
    result = [result, right - left + 1].max
  end

  result
end

File: 1438-longest-continuous-subarray-with-absolute-diff-less-than-or-equal-to-limit/1438-longest-continuous-subarray-with-absolute-diff-less-than-or-equal-to-limit.cpp
class Solution {
public:
    int longestSubarray(vector<int>& nums, int limit) {
        deque<int> max_deque, min_deque;
        int left = 0;
        int result = 0;

        for (int right = 0; right < nums.size(); ++right) {
            int num = nums[right];
            while (!max_deque.empty() && max_deque.back() < num) {
                max_deque.pop_back();
            }
            max_deque.push_back(num);

            while (!min_deque.empty() && min_deque.back() > num) {
                min_deque.pop_back();
            }
            min_deque.push_back(num);

            while (max_deque.front() - min_deque.front() > limit) {
                if (max_deque.front() == nums[left]) {
                    max_deque.pop_front();
                }
                if (min_deque.front() == nums[left]) {
                    min_deque.pop_front();
                }
                ++left;
            }

            result = max(result, right - left + 1);
        }

        return result;
    }

};